### ISSUE TITLE
Feature: Ignore image volumes when running containers

### ISSUE THREAD

**Issue by lucasbasquerotto** [moby/moby#43190]
I would like to request an option (configuration) to ignore volumes defined in images when running containers.

More specifically, I would like that volumes are created and attached to containers only if defined explicitly (for example, using `docker run` with `-v` option, or `docker-compose` with `volumes` for a service).

I find it terribly frustrating that a build option ends up having runtime implications (especially considering that the volumes could persist even after the container is removed).

This causes problems like in the mysql image, with issues like https://github.com/docker-library/mysql/issues/255, asking to remove the VOLUME directive from the dockerfile. In some cases, like in the Oracle image, [it was removed](https://github.com/oracle/docker-images/issues/640#issuecomment-607546834).

There are issues like https://github.com/moby/moby/issues/3465 asking to remove/reset properties from the parent image, like an UNVOLUME directive, but they are probably harder to implement. Furthermore, the user of the image need to inspect each image (including new versions, with the risking of unwanted anonymous volumes being created otherwise) to see if there are volumes defined, which would not be ideal.

I even posted an [SO answer advising against VOLUME in Dockerfiles](https://stackoverflow.com/a/62068396/4850646).

This feature would be enabled by setting a flag (so no breaking changes for those that don't need it). The only thing this flag would do, when enabled, is to ignore any volumes defined in the images (as if there was no VOLUME directive in the Dockerfile). 

It would also to ignore VOLUME defined in base images in the build step too, so an image that extends another would be built as if the base image doesn't have VOLUME defined in it. The VOLUME directive would still work in the same stage in which is defined (this would be useful for performance in some cases, but at the same time it would not implicate negatively the consumers of the image).

I don't know how hard it would be to implement it, but considering that the only thing it does is to ignore volumes from images, it seems pretty straightforward, and for me (and probably many others) it would be very useful.

**Comment by thaJeztah** [comment]
> It would also to ignore VOLUME defined in base images in the build step too, so an image that extends another would be built as if the base image doesn't have VOLUME defined in it. The VOLUME directive would still work in the same stage in which is defined (this would be useful for performance in some cases, but at the same time it would not implicate negatively the consumers of the image).

If I'm not mistaken, this is already the case when using BuildKit as builder (which treats `VOLUME` as "metadata only" during build)

**Comment by lucasbasquerotto** [comment]
@thaJeztah Thanks, that's good to know. Is there some documentation about that? In the docs I have only found (at https://docs.docker.com/engine/reference/builder/#volume) that:

> Changing the volume from within the Dockerfile: If any build steps change the data within the volume after it has been declared, those changes will be discarded.

I guess it's considering without buildkit. I only found a reference to that behaviour in a comment at https://github.com/moby/buildkit/issues/880#issuecomment-473460544 that says:

> Buildkit fixes it by making the VOLUME metadata only-command like the rest so it only affects the final image. If you for some reason want a writable area in your build that doesn't persist you can use the --mount type=tmpfs,target=/path in buildkit

So, it seems that this feature is kind obscure in the documentation, but it's good to know that VOLUME with buildkit doesn't affect the build stage anymore (but you still can create a temporary mount if performance is a concern).

This issue can be about only the runtime behaviour, then.

**Comment by nevets963** [comment]
@lucasbasquerotto I have the same problem. Just wondered if you found any other useful info since Jan? Thanks! Steve

**Comment by nevets963** [comment]
@lucasbasquerotto Thanks for the responding!

Re
> then I change to use an explicit mount point, so as to not create anonymous volumes

Do you mean, if an image has:
```
VOLUME /var/lib/somedata
```

You would simply add an explicit mount like in eg docker-compose: `/host/data:/var/lib/somedata` and this avoids the anomymous volume?

I have no idea if this works, but I was considering whether I could make `/var/lib/docker/volumes` be a `nullfs` file system... But I don't think so as Docker puts `metadata.db` in there and I assume that is important.

**Comment by lucasbasquerotto** [comment]
@nevets963 Yes, directories defined as VOLUME's in a Dockerfile won't create anonymous volumes if they are explicitly mapped to the host.

**Example:**

_Dockerfile:_

```Dockerfile
FROM ubuntu:20.04

VOLUME /var/test
```

Build the image and check the volumes in your machine:

```bash
docker image build -t local/test .
docker volume ls
```

Run a container **without** the `--rm` flag, then remove the container and check that the new anonymous volume created remains:

```bash
docker run local/test echo test
docker container prune -f
docker volume ls
```

Remove the volume (use the volume id you got on the previous step):

```bash
docker volume remove <anonymous_volume_id>
docker volume ls
```

Run a new container using the same image, but specifying a bind mount in the same path of the volume in the dockerfile, and then check that no volume was created:

```bash
docker run -v /var/test:/var/test local/test echo test
docker volume ls
```

> I have no idea if this works, but I was considering whether I could make /var/lib/docker/volumes be a nullfs file system...

I have never tried that, so I don't know.

**Comment by nevets963** [comment]
@lucasbasquerotto Ok, got it! Thanks for the detail, I hope this helps someone else too!

**Comment by polarathene** [comment]
It would probably be good as a daemon config option too for those that always want to default to running containers without implicit anonymous volumes.

There's no drawbacks to running containers this way AFAIK, explicit volumes can be configured when there is an actual need for a distinct volume (persistence/performance).

**Comment by polarathene** [comment]
FWIW, [Podman has the `--image-volume` option](https://docs.podman.io/en/v5.1.1/markdown/podman-run.1.html#image-volume-bind-tmpfs-ignore) that defaults to `bind`, while offering alternatives `tmpfs` for ephermal use and `ignore` for mounting anything at the path entirely. Something similar to that might make sense, along with a `daemon.json` config to support it.

**Comment by LaurentGoderre** [comment]
The way I would see this potentially working is adding a flag to the `docker run` command like `--ignore-anonymous-volumes`

**Comment by kale1d0code** [comment]
There should have been an option to ignore VOLUME directives in Dockerfiles from within the /etc/docker/daemon.json file from the start.

This would solve the problem of 100's of anonymous volumes being created on my host all the time

having a command argument for docker run would be nice however it would fall short in scenarios where you are creating containers in other ways, (portainer docker-compose etc)

**Comment by AB-xdev** [comment]
I also stumbled over this issue but with other context:

When using [``docker container commit ...``](https://docs.docker.com/reference/cli/docker/container/commit/) it's impossible to dump some containers completely/properly because of their use of ``VOLUME``.

I think proving an option like podman has with ``image-volume`` (and was pointed out in the comments above) would help a lot.
